### PR TITLE
Creates index as per data model

### PR DESCRIPTION
**Pull request by Kajal-NEC**
This pull request fixes issue #2204

### CODE REVIEW

**Comment by fgalan** [comment]
The usual DM used with STHSink and MongoSink is `dm-by-service-path`

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java [hunk 1]
```diff
@@ -102,12 +102,26 @@ public void createCollection(String dbName, String collectionName, long dataExpi
...
-                .append("_id.resolution", 1)
-                .append("_id.origin", 1);
+            keys = new BasicDBObject();
+                switch(dataModel) {
+                case DMBYSERVICEPATH:
+                    keys.append("_id.entityId", 1)
+                        .append("_id.entityType", 1)
+                        .append("_id.attrName", 1)
+                        .append("_id.origin", 1)
+                        .append("_id.resolution", 1);
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java:112]: Why this change in the order of keys instead of keeping the original one? I mean:

```suggestion
                        .append("_id.resolution", 1);
                        .append("_id.origin", 1)
```

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java [hunk 2]
```diff
@@ -102,12 +102,26 @@ public void createCollection(String dbName, String collectionName, long dataExpi
...
+                    break;
+                case DMBYENTITY:
+                    keys.append("_id.attrName", 1)
+                        .append("_id.origin", 1)
+                        .append("_id.resolution", 1);
+                    break;
+                case DMBYATTRIBUTE:
+                    keys.append("_id.origin", 1)
+                         .append("_id.resolution", 1);
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java:121]:
```suggestion
                        .append("_id.resolution", 1);
```

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java [hunk 3]
```diff
@@ -102,12 +102,26 @@ public void createCollection(String dbName, String collectionName, long dataExpi
         BasicDBObject keys;
         IndexOptions options;
         try {
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java:104]: From which sink is this `createCollection()` is being called? MongoSink or STHSink?

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java [hunk 4]
```diff
@@ -162,14 +176,33 @@ public void createCollection(String dbName, String collectionName, long collecti
         } // try catch
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java:176]: From which sink is this `createCollection()` is being called? MongoSink or STHSink?

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java (hunk 2 above)
Comment (Kajal-NEC) [reply cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java]: Resolved in f7e74e4

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java (hunk 3 above)
Comment (Kajal-NEC) [reply cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java:104]: `createCollection(String dbName, String collectionName, long dataExpiration)` is being called from [STHSink](https://github.com/telefonicaid/fiware-cygnus/blob/master/cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSISTHSink.java#L376)

**Comment by Kajal-NEC** [comment]
Hi @fgalan ,
I have resolved both the suggestions. Please merge the PR If it seems ok.

**Review by AlvaroVega — APPROVED** [review]
LGTM

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java [hunk 5]
```diff
@@ -102,12 +102,26 @@ public void createCollection(String dbName, String collectionName, long dataExpi
         BasicDBObject keys;
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java:102]: An entry in CHANGES_NEXT_RELEASE file should be included, describing the changes in this PR. Something like this:

```
[cygnus-common] MongoDB indexes are created depending on DM (#2204)
```

Comment (Kajal-NEC) [reply cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackendImpl.java:102]: Added in 4ff6d9b

**Comment by Kajal-NEC** [comment]
Hi @fgalan,
I have added entry in CNR.
Thanks.

**Review by fgalan — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by AlvaroVega** [comment]
Documentation about sth indexes: https://github.com/telefonicaid/fiware-sth-comet/blob/master/doc/manuals/db_indexes.md should be updated after this PR changes then depending on the datamodel used.
